src/zkevm_specs/evm/execution/end_tx.py

```diff
@@ -17,6 +18,10 @@ def end_tx(instruction: Instruction):
     refund = instruction.tx_refund_read(tx_id)
     effective_refund = instruction.min(max_refund, refund, 8)
 
+    # gas_used & refund == 0 if tx is invalid
+    instruction.constrain_zero(is_tx_invalid * gas_used)
+    instruction.constrain_zero(is_tx_invalid * effective_refund)
```

Comment:
Hmm shouldn't this be enforced elsewhere, or maybe just be something like 
```
gas_used = is_tx_invalid * gas_used;
effective_refund = is_tx_invalid * effective_refund;
```

Both `gas_used` and `effective_refund` seem to be calculated directly from other data sources so I don't know how this check helps. Either it's already always `0` because of some other checks in `begin_tx` or something, or this check is useless because `gas_used` and `effective_refund` still have their normal values  when `is_tx_invalid` is true?

Reply:
Could be I think, let me check if they are inherited from begin_tx or not.

Reply:
Could be merged into gas handling logic in begin_tx, which makes gas_left equal tx.gas if tx is invalid, so that here gas_used == 0.

Reply:
Hmmm yes seems like refund is read from the tx table but not checked anywhere else. If that's the case the new code makes sense to me.